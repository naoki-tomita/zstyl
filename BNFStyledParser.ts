type Types =
    "StyleSheet"
  | "LocalStyle"
  | "Style"
  | "Identifier"
  | "Value"
  | "Values"
  | "NestedStyle"
  | "Selector"
  | "Block"
  | "MediaCondition"
  | "Media"
  | "Keyframe"
  ;

interface Ast {
  type: Types;
};

type RootStyle = LocalStyle | NestedStyle | MediaStyle | KeyframeStyle;

export interface StyleSheetAst extends Ast {
  type: "StyleSheet";
  children: Array<LocalStyle | NestedStyle | MediaStyle | KeyframeStyle>;
}

export const StyleSheetParser: Parser<StyleSheetAst> = {
  parse(style: string): ParsedResult<StyleSheetAst> {
    const children: StyleSheetAst["children"] = [];
    let result: ParsedResult<RootStyle> = { remaining: style };
    let successToParse;
    do {
      successToParse = false;
      for (const parser of [LocalStyleParser, NestedStyleParser, MediaStyleParser, KeyframeStyleParser]) {
        result = parser.parse(result.remaining);
        if (result.ast) {
          children.push(result.ast);
          successToParse = true;
        }
      }
    } while (successToParse)

    return {
      ast: {
        type: "StyleSheet",
        children,
      },
      remaining: result.remaining
    };
  }
}

export interface Identifier extends Ast {
  type: "Identifier";
  name: string;
}

export const IdentifierParser: Parser<Identifier> = {
  parse(style: string): ParsedResult<Identifier> {
    style = style.trimStart();
    const result = style.match(/^([a-zA-Z0-9\-_.]+)/);
    if (result == null || result[0].length === 0) {
      return { remaining: style };
    }
    const name = result[0];
    return {
      ast: {
        type: "Identifier",
        name,
      },
      remaining: style.slice(name.length),
    }
  }
}

export interface Value extends Ast {
  type: "Value";
  value: string;
}

export const ValueParser: Parser<Value> = {
  parse(style: string): ParsedResult<Value> {
    const result = style.trimStart().match(/^([a-zA-Z0-9\-%#_.\()]+)/);
    if (result == null || result[0].length === 0) {
      return { remaining: style };
    }
    const value = result[0];
    return {
      ast: {
        type: "Value",
        value,
      },
      remaining: style.trimStart().slice(value.length),
    }
  }
}

export interface Values extends Ast {
  type: "Values";
  values: Value[];
}

export const ValuesParser: Parser<Values> = {
  parse(style: string): ParsedResult<Values> {
    let result = ValueParser.parse(style);
    if (!result.ast) {
      return { remaining: style };
    }
    const results: Value[] = [result.ast];
    while (true) {
      const remaining = result.remaining;
      result = ValueParser.parse(remaining);
      if (!result.ast) {
        return {
          ast: {
            type: "Values",
            values: results,
          },
          remaining,
        }
      }
      results.push(result.ast);
    }
  }
}

export interface Style extends Ast {
  type: "Style"
  prop: Identifier;
  values: Values;
}
export const StyleParser: Parser<Style> = {
  parse(style: string): ParsedResult<Style> {
    const identifier = IdentifierParser.parse(style);
    if (!identifier.ast) {
      return { remaining: style };
    }
    const colonWithIdentifiers = identifier.remaining.trimStart();
    if (!colonWithIdentifiers.startsWith(":")) {
      return { remaining: style };
    }
    const identifiers = ValuesParser.parse(colonWithIdentifiers.slice(1));
    if (!identifiers.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "Style",
        prop: identifier.ast,
        values: identifiers.ast,
      },
      remaining: identifiers.remaining
    }
  }
}

export interface LocalStyle extends Ast {
  type: "LocalStyle";
  prop: Identifier;
  values: Values;
}
export const LocalStyleParser: Parser<LocalStyle> = {
  parse(style: string): ParsedResult<LocalStyle> {
    const result = StyleParser.parse(style);
    if (
      !result.ast ||
      !result.remaining.trimStart().startsWith(";")
    ) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "LocalStyle",
        prop: result.ast.prop,
        values: result.ast.values,
      },
      remaining: result.remaining.trimStart().slice(1)
    }
  }
}

type AnyStyle = LocalStyle | NestedStyle;

export interface Selector extends Ast {
  type: "Selector";
  value: string; // ここを BNF で表現するのはあきらめました
}

export const SelectorParser: Parser<Selector> = {
  parse(style: string): ParsedResult<Selector> {
    const result = style.trimStart().match(/^([a-zA-Z0-9_()+>| ,.#~=%^$&\[\]"'*:/\-]+)/)
    if (result == null) {
      return { remaining: style };
    }
    const value = result[0].trim()
    return {
      ast: {
        type: "Selector",
        value,
      },
      remaining: style.trimStart().slice(value.length)
    }
  }
}

export interface Block extends Ast {
  type: "Block";
  body: AnyStyle[];
}

export const BlockParser: Parser<Block> = {
  parse(style: string): ParsedResult<Block> {
    if (!style.trimStart().startsWith("{")) {
      return { remaining: style };
    }
    let result: ParsedResult<AnyStyle> = { remaining: style.trimStart().slice(1) };
    const results: AnyStyle[] = [];
    let successToParse: boolean;
    do {
      successToParse = false;
      for (const parser of [LocalStyleParser, NestedStyleParser]) {
        result = parser.parse(result.remaining);
        if (result.ast) {
          successToParse = true;
          results.push(result.ast);
        }
      }
    } while (successToParse)
    if (!result.remaining.trimStart().startsWith("}")) {
      return { remaining: style }
    }
    return {
      ast: {
        type: "Block",
        body: results,
      },
      remaining: result.remaining.trimStart().slice(1)
    }
  }
}

export interface NestedStyle extends Ast {
  type: "NestedStyle";
  selector: Selector;
  block: Block;
}

export const NestedStyleParser: Parser<NestedStyle> = {
  parse(style: string): ParsedResult<NestedStyle> {
    const selector = SelectorParser.parse(style);
    if (!selector.ast) {
      return { remaining: style };
    }
    const block = BlockParser.parse(selector.remaining);
    if (!block.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "NestedStyle",
        selector: selector.ast,
        block: block.ast,
      },
      remaining: block.remaining
    }
  }
}

interface MediaCondition extends Ast {
  type: "MediaCondition";
  name: Identifier;
  value: Identifier;
}

export interface MediaStyle extends Ast {
  type: "Media"
  condition: MediaCondition;
  block: Block;
}

export const MediaStyleParser: Parser<MediaStyle> = {
  parse(style: string): ParsedResult<MediaStyle> {
    if (!style.trimStart().startsWith("@media")) {
      return { remaining: style };
    }
    const slicedAtMediaStyle = style.trimStart().slice("@media".length);
    if (!slicedAtMediaStyle.trimStart().startsWith("(")) {
      return { remaining: style };
    }
    const leftParenTrimmedStyle = slicedAtMediaStyle.trimStart().slice(1);
    const nameResult = IdentifierParser.parse(leftParenTrimmedStyle);
    if (!nameResult.ast) {
      return { remaining: style };
    }
    if (!nameResult.remaining.trimStart().startsWith(":")) {
      return { remaining: style };
    }
    const valueResult = IdentifierParser.parse(nameResult.remaining.trimStart().slice(1));
    if (!valueResult.ast) {
      return { remaining: style };
    }
    if (!valueResult.remaining.trimStart().startsWith(")")) {
      return { remaining: style };
    }
    const rightParenTrimmedStyle = valueResult.remaining.trimStart().slice(1);
    const blockResult = BlockParser.parse(rightParenTrimmedStyle);
    if (!blockResult.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "Media",
        condition: {
          type: "MediaCondition",
          name: nameResult.ast,
          value: valueResult.ast,
        },
        block: blockResult.ast
      },
      remaining: blockResult.remaining
    }
  }
}

export interface KeyframeStyle extends Ast {
  type: "Keyframe"
  name: Identifier;
  block: Block;
}

export const KeyframeStyleParser: Parser<KeyframeStyle> = {
  parse(style: string): ParsedResult<KeyframeStyle> {
    if (!style.trimStart().startsWith("@keyframe")) {
      return { remaining: style };
    }
    const keyframeExcludedStyle = style.trimStart().slice("@keyframe".length);
    const nameResult = IdentifierParser.parse(keyframeExcludedStyle);
    if (!nameResult.ast) {
      return { remaining: style };
    }
    const blockStyle = nameResult.remaining;
    const blockResult = BlockParser.parse(blockStyle);
    if (!blockResult.ast) {
      return { remaining: style };
    }
    return {
      ast: {
        type: "Keyframe",
        name: nameResult.ast,
        block: blockResult.ast
      },
      remaining: blockResult.remaining,
    }
  }
}

type ParsedResult<T extends Ast> = {
  ast?: T, remaining: string
};

interface Parser<T extends Ast> {
  parse(style: string): ParsedResult<T>;
}
