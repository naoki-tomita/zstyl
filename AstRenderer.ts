import type { Block, Identifier, Values, KeyframeStyle, LocalStyle, MediaStyle, NestedStyle, StyleSheetAst, Value } from "./BNFStyledParser"
import { StyleSheetParser } from "./BNFStyledParser";

function space(width: number) {
  return Array(width).fill(null).map(() => " ").join("");
}
function indent(text: string, width: number) {
  return text.split("\n").map(t => `${space(width)}${t}`).join(`\n`);
}
export const AstRenderer = {
  renderStyleSheetWithId(id: string, stylesheet: StyleSheetAst): string {
    const localStyles = stylesheet.children.filter((it): it is LocalStyle => it.type === "LocalStyle");
    const others = stylesheet.children.filter((it): it is (NestedStyle | MediaStyle | KeyframeStyle) => it.type !== "LocalStyle");
    const localStylesResult = this.renderLocalStylesWithId(id, localStyles);
    const othersResults = others.map(style => {
      switch (style.type) {
        case "NestedStyle":
          return this.renderNestedStyleWithId(id, style);
        case "Media":
          return this.renderMediaStyleWithId(id, style);
        case "Keyframe":
          return this.renderKeyframeStyleWithId(id, style);
      }
    });
    return [localStylesResult, ...othersResults].join("\n");
  },

  renderLocalStylesWithId(id: string, localStyles: LocalStyle[]): string {
    if (localStyles.length === 0) {
      return ``;
    }
    return `${id} { ${
      localStyles
        .map(it => `${this.renderIdentifier(it.prop)}: ${this.renderIdentifiers(it.values)};`)
        .join(" ")
    } }`;
  },

  renderNestedStyleWithId(id: string, nestedStyle: NestedStyle): string {
    const newId = nestedStyle.selector.value.includes("&")
      ? nestedStyle.selector.value.replaceAll("&", id)
      : `${id} ${nestedStyle.selector.value}`;
    return this.renderBlock(newId, nestedStyle.block);
  },

  renderMediaStyleWithId(id: string, mediaStyle: MediaStyle): string {
    return `@media (${
      this.renderIdentifier(mediaStyle.condition.name)
    }: ${
      this.renderIdentifier(mediaStyle.condition.value)
    }) {
${indent(this.renderBlock(id, mediaStyle.block), 2)}
}`;
  },

  renderKeyframeStyleWithId(id: string, mediaStyle: KeyframeStyle): string {
    return `@keyframes ${this.renderIdentifier(mediaStyle.name)} {
${indent(this.renderBlock("", mediaStyle.block), 2)}
}`;
  },

  renderBlock(id: string, block: Block) {
    const localStyles = block.body.filter((it): it is LocalStyle => it.type === "LocalStyle");
    const nestedStyles = block.body.filter((it): it is NestedStyle => it.type === "NestedStyle");
    const localStylesResult = this.renderLocalStylesWithId(id, localStyles);
    const nestedStyleResults = nestedStyles.map(style => this.renderNestedStyleWithId(id, style));
    return [localStylesResult, ...nestedStyleResults].join("\n").trimStart();
  },

  renderIdentifiers(identifiers: Values): string {
    return identifiers.values.map(this.renderValue).join(" ");
  },

  renderValue(value: Value): string {
    return value.value;
  },

  renderIdentifier(identifier: Identifier): string {
    return identifier.name;
  },
}
