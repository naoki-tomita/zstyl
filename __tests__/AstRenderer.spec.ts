import { StyleSheetParser } from "../BNFStyledParser";
import { AstRenderer } from "../AstRenderer";

describe("AstRenderer", () => {
  describe('fillWithId', () => {
    it("should parse style and fill id.", () => {
      const style = `
        display: flex;

        &:hover {
          color: red;
        }

        .wrapper {
          color: blue;
        }

        @media (max-width: 200px) {
          display: inline-flex;
        }

        @keyframes keyframe-name {
          0% {
            height: 10px;
          }

          100% {
            height: 100px;
          }
        }
      `;
      const stylesheet = StyleSheetParser.parse(style);
      expect(AstRenderer.renderStyleSheetWithId("#element", stylesheet.ast!)).toMatchSnapshot();
    });

    it("should parse and fill id for multiple nested style.", () => {
      const style = `
        display: flex;

        &:hover {
          color: red;

          .inner {
            color: black;
          }
        }

        .wrapper {
          color: blue;

          .inner {
            color: green;
          }
        }

        @media (max-width: 200px) {
          display: inline-flex;

          .inner {
            color: yellow;
          }
        }
      `;

      const stylesheet = StyleSheetParser.parse(style);
      expect(AstRenderer.renderStyleSheetWithId("#element", stylesheet.ast!)).toMatchSnapshot();
    });

    it("should real pattern", () => {
      const style = `
        display: flex;
        gap: 16px;
        justify-content: center;

        @media (max-width: 940px) {
          flex-direction: column;
        }

        @keyframes keyframe-name {
          0% {
            height: 10px;
          }

          100% {
            height: 100px;
          }
        }
      `;
      const stylesheet = StyleSheetParser.parse(style);
      expect(AstRenderer.renderStyleSheetWithId("*[data-zstyl='aaaaa']", stylesheet.ast!)).toMatchSnapshot();
    });
  });

});
