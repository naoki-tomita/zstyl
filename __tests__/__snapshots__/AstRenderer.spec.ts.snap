// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AstRenderer fillWithId should parse and fill id for multiple nested style. 1`] = `
"#element { display: flex; }
#element:hover { color: red; }
#element:hover .inner { color: black; }
#element .wrapper { color: blue; }
#element .wrapper .inner { color: green; }
@media (max-width: 200px) {
  #element { display: inline-flex; }
  #element .inner { color: yellow; }
}"
`;

exports[`AstRenderer fillWithId should parse style and fill id. 1`] = `
"#element { display: flex; }
#element:hover { color: red; }
#element .wrapper { color: blue; }
@media (max-width: 200px) {
  #element { display: inline-flex; }
}
@keyframes keyframe-name {
  0% { height: 10px; }
  100% { height: 100px; }
}"
`;

exports[`AstRenderer fillWithId should real pattern 1`] = `
"*[data-zstyl='aaaaa'] { display: flex; gap: 16px; justify-content: center; }
@media (max-width: 940px) {
  *[data-zstyl='aaaaa'] { flex-direction: column; }
}
@keyframes keyframe-name {
  0% { height: 10px; }
  100% { height: 100px; }
}"
`;
